Dev/2p ecdsa ios #27

vhnatyk opened the pull request:
Reworked pull request that covers all initial questions

* Ability to call locally at least one function of 2-party ECDSA on iOS:
   all functions from 2-party ECDSA compile and run on iOS 👍 
* Documentation on how to run:
   added wiki page covering whole repo
* Simple benches (execution time for functions of your choice):
   **31.889s**  for **bench_full_keygen_party_one_two** on **iPhone 6S** (comparing to 4.0988s on Mac Mini i5 16GB RAM)

omershlo commented:
@vhnatyk as usual great work. really interested to what functions/operations take most of the running time. Maybe we can learn from it and improve

omershlo commented on Cargo.toml:
```diff
@@ -21,7 +21,7 @@ license = "GPL-3.0-or-later"
 
 [lib]
-crate-type = ["lib"]
+crate-type = ["staticlib", "rlib"]
```
why this change is needed? what is the implication?

vhnatyk replied:
iOS must be build with staticlib as too many dependencies would be missing, and x86_64 host should be fine anyway (faster static libs with slower loading times due to a bigger size vs smaller but slower dynamic libs with faster loading times)  but... I just realized all platform-specific overrides can be placed in dinghy.toml!! - will fix that asap

vhnatyk replied:
moved this from Cargo.toml to dinghy.toml, don't see why GitHub not showing latest changes :|

omershlo commented on Cargo.toml:
```diff
@@ -34,7 +34,7 @@ reqwest = "0.9.5"
 rocket_contrib = "0.4.0"
 uuid = { version = "0.7", features = ["v4"] }
-rust-crypto = "^0.2"
+rust-crypto = { git = "https://github.com/vhnatyk/rust-crypto", branch="aarch64"} #"^0.2"
```
what is the change between the "official" rust-crypto and yours? 
lets talk about this point because it is problematic in terms of upgradability

vhnatyk replied:
I forked where author implemented method `rust_crypto_util_fixed_time_eq_asm` that was missing in original [rust-crypto](https://github.com/DaGenix/rust-crypto) for arm64 - as rust-crypto is not actively developed (perhaps due to being quite feature-complete?), those changes were not merged to master. The method is not used neither in core of gg18, not in core of lindell2017, but in benches for lindell2017 - so it's an open question is it required. I had to fork it,  but maybe it's a good idea for KZen to fork it as rust-crypto isn't actively changing anymore anyway (just like you forked some other libs). The only chages from original `rust-crypto` **master** is that it `Adds a definition of rust_crypto_util_fixed_time_eq_asm for ARMv8. It's
exactly the same as for ARMv7, except that it uses ARMv8 names for
registers (wN instead of rN)` Looks to me like there are no good or obvious reasons why it wasn't merged to master of rust-crypto. @omershlo What do you think?

omershlo replied:
I use rust_crypto only for GG18 with network (in the bin folder). It is suppose to be possible to run the test without this library...
having said that - did you send a PR for rust_crypto?
If we couldn't make it go away we will probably fork it as you suggest.

vhnatyk replied:
Original PR from the branch that I forked is hanging there for quite a while(few years) - no point to do another PR with same changes I think, as rust_crypto seems not maintained anymore (looks like it was split into separate libraries that got abandoned too). Let me rethink a bit with this, I'm working with Android right now, and the changes don't compile there ok with GCC so I'm trying clang - but I found another repo, where author fixed for both iOS and Android for aarch64 and it compiles ok for both platforms. At the moment I'm analyzing those changes

vhnatyk replied:
Imho fork by KZen is the best option - changes for compatibility with aarch64 to rust_crypto_util_fixed_time_eq_asm are relatively minor and seem to look solid, hmm

vhnatyk commented:
> is this really necessary to include bin/gmp_ios.zip? what is being used for
> same question but about the .rtf files in doc.

it's not necessary for sure, but makes it much much easier to run on mobile for somebody new to Rust, GMP and cross-compilation. I added it for the sole purpose of making it easier to recreate. In terms of licensing, GMP can be distributed either in GPL or LGPL  - both licensing models fit licensing model of KZen repo. But it's purely optional. Same applies to big rtf log file - probably can be replaced with cool fancy gifs 👍  that are pretty common ([Dinghy](https://github.com/snipsco/dinghy) also has demo gif btw)

vhnatyk commented:
> > is this really necessary to include bin/gmp_ios.zip? what is being used for
> > same question but about the .rtf files in doc.
> 
> it's not necessary for sure, but makes it much much easier to run on mobile for somebody new to Rust, GMP and cross-compilation. I added it for the sole purpose of making it easier to recreate. In terms of licensing, GMP can be distributed either in GPL or LGPL - both licensing models fit licensing model of KZen repo. But it's purely optional. Same applies to big rtf log file - probably can be replaced with cool fancy gifs 👍 that are pretty common ([Dinghy](https://github.com/snipsco/dinghy) also has demo gif btw)

removed the files (and cleaned up wiki)
